File: kotlin-typescript/karakum/plugins/convertWithMetadata.js
import ts from "typescript"
import * as karakum from "karakum"

export default function (node, context, render) {
    if (
        ts.isTypeAliasDeclaration(node)
        && node.name.text === "WithMetadata"
        && ts.isIntersectionTypeNode(node.type)
        && node.type.types.length > 1
        && ts.isTypeLiteralNode(node.type.types[1])
    ) {
        const inheritanceModifierService = context.lookupService(karakum.inheritanceModifierServiceKey)

        const inheritanceModifier = inheritanceModifierService?.resolveInheritanceModifier(node, context)

        const typeParameters = node.typeParameters
            ?.map(typeParameter => render(typeParameter))
            ?.filter(Boolean)
            ?.join(", ")

        const members = node.type.types[1].members
            .map(member => render(member))
            .join("\n")

        return `
${karakum.ifPresent(inheritanceModifier, it => `${it} `)}external interface WithMetadata${karakum.ifPresent(typeParameters, it => `<${it}>`)} {
${members}

@Suppress(
    "WRONG_BODY_OF_EXTERNAL_DECLARATION",
    "INLINE_EXTERNAL_DECLARATION",
    "NON_ABSTRACT_MEMBER_OF_EXTERNAL_INTERFACE",
    "DECLARATION_CANT_BE_INLINED",
)
inline val value: T
    get() = unsafeCast<T>()
}
        `
    }

    return null
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/typescript/karakum/plugins/convertWithMetadata.kt
package typescript.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.ifPresent
import io.github.sgrishchenko.karakum.extension.plugins.InheritanceModifierService
import io.github.sgrishchenko.karakum.extension.plugins.inheritanceModifierServiceKey
import typescript.asArray
import typescript.isIntersectionTypeNode
import typescript.isTypeAliasDeclaration
import typescript.isTypeLiteralNode

val convertWithMetadata = createPlugin { node, context, render ->
    nullable {
        ensure(isTypeAliasDeclaration(node))
        ensure(node.name.text == "WithMetadata")

        val intersection = node.type
        ensure(isIntersectionTypeNode(intersection))
        ensure(intersection.types.asArray().size > 1)

        val secondType = intersection.types.asArray()[1]
        ensure(isTypeLiteralNode(secondType))

        val inheritanceModifierService =
            ensureNotNull(context.lookupService(inheritanceModifierServiceKey))

        val inheritanceModifier = inheritanceModifierService.resolveInheritanceModifier(node, context)

        val typeParameters = node.typeParameters?.asArray()
            ?.map { render(it) }
            ?.filter { it.isNotEmpty() }
            ?.joinToString(", ")

        val members = secondType.members.asArray().joinToString("\n") { render(it) }

        """
            ${ifPresent(inheritanceModifier) { "$it "}}external interface WithMetadata${ifPresent(typeParameters) { "<$it>"}} {
            $members

            @Suppress(
                "WRONG_BODY_OF_EXTERNAL_DECLARATION",
                "INLINE_EXTERNAL_DECLARATION",
                "NON_ABSTRACT_MEMBER_OF_EXTERNAL_INTERFACE",
                "DECLARATION_CANT_BE_INLINED",
            )
            inline val value: T
                get() = unsafeCast<T>()
            }
        """.trimIndent()
    }
}
